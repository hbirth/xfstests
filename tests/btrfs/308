#! /bin/bash
# SPDX-License-Identifier: GPL-2.0
# Copyright (c) 2023 Western Digital Cooperation.  All Rights Reserved.
#
# FS QA Test 308
#
# Test on-disk layout of RAID Stripe Tree Metadata by writing 128k to an empty
# file on a filesystem that has one stripe already pre-filled. Afterwards
# overwrite a portion of the file.
#
. ./common/preamble
_begin_fstest auto quick raid remount volume raid-stripe-tree

. ./common/filter
. ./common/filter.btrfs

_require_btrfs_command inspect-internal dump-tree
_require_btrfs_mkfs_feature "raid-stripe-tree"
_require_scratch_dev_pool 4
_require_btrfs_fs_feature "raid_stripe_tree"
_require_btrfs_fs_feature "free_space_tree"
_require_btrfs_free_space_tree
_require_btrfs_no_compress
_require_btrfs_no_nodatacow

test $(_get_page_size) -eq 4096 || _notrun "this tests requires 4k pagesize"

test_128k_write_overwrite()
{
	local profile=$1
	local ndevs=$2

	_scratch_dev_pool_get $ndevs

	echo "==== Testing $profile ===="
	_scratch_pool_mkfs -d $profile -m $profile -O raid-stripe-tree
	_scratch_mount

	$XFS_IO_PROG -fc "pwrite -W 0 32k" "$SCRATCH_MNT/bar" | _filter_xfs_io
	$XFS_IO_PROG -fc "pwrite -W 0 128k" "$SCRATCH_MNT/foo" | _filter_xfs_io
	$XFS_IO_PROG -fc "pwrite -W 64k 8k" "$SCRATCH_MNT/foo" | _filter_xfs_io

	_scratch_cycle_mount
	md5sum "$SCRATCH_MNT/foo" | _filter_scratch

	_scratch_unmount

	$BTRFS_UTIL_PROG inspect-internal dump-tree -t raid_stripe $SCRATCH_DEV_POOL |\
		_filter_stripe_tree

	_scratch_dev_pool_put
}

echo "= Test 128k write to empty file with 1st stripe partially prefilled then overwrite ="
test_128k_write_overwrite raid0 2
test_128k_write_overwrite raid1 2
test_128k_write_overwrite raid10 4

# success, all done
status=0
exit
