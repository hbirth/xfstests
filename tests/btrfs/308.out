QA output created by 308
= Test 128k write to empty file with 1st stripe partially prefilled then overwrite =
==== Testing raid0 ====
wrote 32768/32768 bytes at offset 0
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
wrote 131072/131072 bytes at offset 0
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
wrote 8192/8192 bytes at offset 65536
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
d48858312a922db7eb86377f638dbc9f  SCRATCH_MNT/foo

raid stripe tree key (RAID_STRIPE_TREE ROOT_ITEM 0)
leaf XXXXXXXXX items X free space XXXXX generation X owner RAID_STRIPE_TREE
leaf XXXXXXXXX flags 0x1(WRITTEN) backref revision 1
checksum stored <CHECKSUM>
checksum calced <CHECKSUM>
fs uuid <UUID>
chunk uuid <UUID>
	item 0 key (XXXXXX RAID_STRIPE 32768) itemoff XXXXX itemsize 24
			encoding: RAID0
			stripe 0 devid 1 physical XXXXXXXXX
	item 1 key (XXXXXX RAID_STRIPE 32768) itemoff XXXXX itemsize 24
			encoding: RAID0
			stripe 0 devid 1 physical XXXXXXXXX
	item 2 key (XXXXXX RAID_STRIPE 65536) itemoff XXXXX itemsize 24
			encoding: RAID0
			stripe 0 devid 2 physical XXXXXXXXX
	item 3 key (XXXXXX RAID_STRIPE 32768) itemoff XXXXX itemsize 24
			encoding: RAID0
			stripe 0 devid 1 physical XXXXXXXXX
	item 4 key (XXXXXX RAID_STRIPE 8192) itemoff XXXXX itemsize 24
			encoding: RAID0
			stripe 0 devid 1 physical XXXXXXXXX
total bytes XXXXXXXX
bytes used XXXXXX
uuid <UUID>
==== Testing raid1 ====
wrote 32768/32768 bytes at offset 0
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
wrote 131072/131072 bytes at offset 0
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
wrote 8192/8192 bytes at offset 65536
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
d48858312a922db7eb86377f638dbc9f  SCRATCH_MNT/foo

raid stripe tree key (RAID_STRIPE_TREE ROOT_ITEM 0)
leaf XXXXXXXXX items X free space XXXXX generation X owner RAID_STRIPE_TREE
leaf XXXXXXXXX flags 0x1(WRITTEN) backref revision 1
checksum stored <CHECKSUM>
checksum calced <CHECKSUM>
fs uuid <UUID>
chunk uuid <UUID>
	item 0 key (XXXXXX RAID_STRIPE 32768) itemoff XXXXX itemsize 40
			encoding: RAID1
			stripe 0 devid 1 physical XXXXXXXXX
			stripe 1 devid 2 physical XXXXXXXXX
	item 1 key (XXXXXX RAID_STRIPE 131072) itemoff XXXXX itemsize 40
			encoding: RAID1
			stripe 0 devid 1 physical XXXXXXXXX
			stripe 1 devid 2 physical XXXXXXXXX
	item 2 key (XXXXXX RAID_STRIPE 8192) itemoff XXXXX itemsize 40
			encoding: RAID1
			stripe 0 devid 1 physical XXXXXXXXX
			stripe 1 devid 2 physical XXXXXXXXX
total bytes XXXXXXXX
bytes used XXXXXX
uuid <UUID>
==== Testing raid10 ====
wrote 32768/32768 bytes at offset 0
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
wrote 131072/131072 bytes at offset 0
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
wrote 8192/8192 bytes at offset 65536
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
d48858312a922db7eb86377f638dbc9f  SCRATCH_MNT/foo

raid stripe tree key (RAID_STRIPE_TREE ROOT_ITEM 0)
leaf XXXXXXXXX items X free space XXXXX generation X owner RAID_STRIPE_TREE
leaf XXXXXXXXX flags 0x1(WRITTEN) backref revision 1
checksum stored <CHECKSUM>
checksum calced <CHECKSUM>
fs uuid <UUID>
chunk uuid <UUID>
	item 0 key (XXXXXX RAID_STRIPE 32768) itemoff XXXXX itemsize 40
			encoding: RAID10
			stripe 0 devid 1 physical XXXXXXXXX
			stripe 1 devid 2 physical XXXXXXXXX
	item 1 key (XXXXXX RAID_STRIPE 32768) itemoff XXXXX itemsize 40
			encoding: RAID10
			stripe 0 devid 1 physical XXXXXXXXX
			stripe 1 devid 2 physical XXXXXXXXX
	item 2 key (XXXXXX RAID_STRIPE 65536) itemoff XXXXX itemsize 40
			encoding: RAID10
			stripe 0 devid 3 physical XXXXXXXXX
			stripe 1 devid 4 physical XXXXXXXXX
	item 3 key (XXXXXX RAID_STRIPE 32768) itemoff XXXXX itemsize 40
			encoding: RAID10
			stripe 0 devid 1 physical XXXXXXXXX
			stripe 1 devid 2 physical XXXXXXXXX
	item 4 key (XXXXXX RAID_STRIPE 8192) itemoff XXXXX itemsize 40
			encoding: RAID10
			stripe 0 devid 1 physical XXXXXXXXX
			stripe 1 devid 2 physical XXXXXXXXX
total bytes XXXXXXXX
bytes used XXXXXX
uuid <UUID>
